test/self-replicate.js

```diff
@@ -0,0 +1,170 @@
+const tape = require('tape')
+const crypto = require('crypto')
+const SecretStack = require('secret-stack')
+const { promisify: pify } = require('util')
+
+const u = require('./misc/util')
+
+const sleep = pify(setTimeout)
+const caps = {
+  shs: crypto.randomBytes(32).toString('base64'),
+}
+
+const CONNECTION_TIMEOUT = 500 // ms
+const REPLICATION_TIMEOUT = 2 * CONNECTION_TIMEOUT
+
+function Server(name, opts = {}) {
+  const stack = SecretStack({ caps }).use(require('ssb-db')).use(require('../'))
+
+  return stack({
+    temp: `test-self-replicate-${name}`, // ssb-db only
+    timeout: CONNECTION_TIMEOUT,
+    keys: u.keysFor(name),
+    ...opts,
+  })
+}
+
+tape('alice restores from pub', async (t) => {
+  const aliceKeys = u.keysFor('alice')
+
+  const alice = Server('alice', { keys: aliceKeys })
+  const pub = Server('pub')
+
+  await pify(alice.publish)({ type: 'post', text: 'hello!' })
+  await pify(alice.publish)({ type: 'post', text: 'hello again' })
+  await pify(pub.publish)({ type: 'pub', text: 'i am pub' }) // needed to provide clock info
+
+  t.deepEqual(
+    alice.ebt.peerStatus(alice.id),
+    {
+      id: alice.id,
+      seq: 2,
+      peers: {},
+    },
+    'alice: correct local state'
+  )
+  t.deepEqual(
+    pub.ebt.peerStatus(alice.id),
+    {
+      id: alice.id,
+      seq: undefined,
+      peers: {},
+    },
+    'pub: has nothing on alice'
+  )
+
+  // pub set up to want alice
+  pub.ebt.request(alice.id, true)
+  // pub.ebt.request(pub.id, true) // NOT needed
+  // alice.ebt.request(pub.id, true) // NOT needed
+  alice.ebt.request(alice.id, true) // :fire: MYSTERY - replication fails without this
```

Comment:
At this point in the test we are just trying to get the **pub** to replicate **alice**

Mystery: If we tell the pub it wants alice's feed (ebt.request) then connect alice to the pub, NOTHING HAPPENS (i.e. the pub does not replicate alice).

"solution": if we have alice request herself before the connection, the replication to the pub works. This is **very** counter-intuitive. Seems like a bug not a feature.

Reply:
Alice needs to put herself in the list of things she replicates, otherwise when the pub asks for it, she's not on her list of things to replicate.

It's a bit counter intuitive but ebt has always worked like this and it's consistent internally.